pkg/mountpoint/args.go

```diff
@@ -0,0 +1,144 @@
+package mountpoint
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	ArgForeground      = "--foreground"
+	ArgReadOnly        = "--read-only"
+	ArgAllowOther      = "--allow-other"
+	ArgAllowRoot       = "--allow-root"
+	ArgRegion          = "--region"
+	ArgCache           = "--cache"
+	ArgUserAgentPrefix = "--user-agent-prefix"
+	ArgAWSMaxAttempts  = "--aws-max-attempts"
+)
+
+// An ArgKey represents the key of an argument.
+type ArgKey = string
+
+// An ArgValue represents the value of an argument.
+type ArgValue = string
+
+// A value to use in arguments without any value, i.e., an option.
+const ArgNoValue = ""
+
+// An Arg represents an argument to be passed to Mountpoint.
+type Arg struct {
+	key   ArgKey
+	value ArgValue
+}
+
+// String returns string representation of the argument to pass Mountpoint.
+func (a *Arg) String() string {
+	if a.value == ArgNoValue {
+		return a.key
+	}
+	return fmt.Sprintf("%s=%s", a.key, a.value)
+}
+
+// An Args represents arguments to be passed to Mountpoint during mount.
+type Args struct {
+	args sets.Set[Arg]
+}
+
+// ParseArgs parses given list of unnormalized and returns a normalized [Args].
+func ParseArgs(passedArgs []string) Args {
+	args := sets.New[Arg]()
+
+	for _, arg := range passedArgs {
+		var key, value string
+
+		parts := strings.SplitN(strings.Trim(arg, " "), "=", 2)
+		if len(parts) == 2 {
+			// Ex: --key=value, key=value
+			key, value = parts[0], parts[1]
+		} else {
+			// Ex: --key value, key value
+			// Ex: --key, key
+			parts = strings.SplitN(strings.Trim(parts[0], " "), " ", 2)
+			if len(parts) == 1 {
+				// Ex: --key, key
+				key = parts[0]
+			} else {
+				// Ex: --key value, key value
+				key, value = parts[0], strings.Trim(parts[1], " ")
+			}
+		}
+
+		// prepend -- if it's not already there
+		key = normalizeKey(key)
+
+		// disallow options that don't make sense in CSI
+		switch key {
+		case "--foreground", "-f", "--help", "-h", "--version", "-v":
+			continue
+		}
+
+		args.Insert(Arg{key, value})
+	}
+
+	return Args{args}
+}
+
+// Set sets or replaces value of given key.
+func (a *Args) Set(key ArgKey, value ArgValue) {
+	key = normalizeKey(key)
+	a.Remove(key)
+	a.args.Insert(Arg{key, value})
+}
+
+// Value extracts value of given key, it returns extracted value and whether the key was found.
+func (a *Args) Value(key ArgKey) (ArgValue, bool) {
+	arg, exists := a.find(key)
+	return arg.value, exists
+}
+
+// Has returns whether given key exists in [Args].
+func (a *Args) Has(key ArgKey) bool {
+	_, exists := a.find(key)
+	return exists
+}
+
+// Remove removes given key, it returns the key's value and whether the key was found.
+func (a *Args) Remove(key ArgKey) (ArgValue, bool) {
+	arg, exists := a.find(key)
+	if exists {
+		a.args.Delete(arg)
+	}
+	return arg.value, exists
+}
+
+// SortedList returns ordered list of normalized arguments.
+func (a *Args) SortedList() []string {
```

Comment:
Nit: Why do we need this? If it's just for equality, couldn't we return a set? (Presumably this isn't called enough to care about time complexity though)

Reply:
It's used while passing these arguments to Mountpoint via [systemd](https://github.com/awslabs/mountpoint-s3-csi-driver/pull/349/files#diff-7d42d437046be1e45cd3cd87a2685779f91fce6ddd67d6d097d1fe395f7fb8cdR158) or via [exec.Command](https://github.com/awslabs/mountpoint-s3-csi-driver/pull/349/files#diff-69831569493f68f24f15945836d12837d45e5a5099bbf0e397001c4b6c91861eR65) and also in tests as you mentioned